File: 13-sorting/5-custom-comparators/main.cpp
// Custom Comparator - C++ Example

#include <iostream>
#include <vector>
#include <algorithm>

struct Student {
    std::string name;
    int grade;
};

bool cmp(const Student& a, const Student& b) {
    if (a.grade != b.grade)
        return a.grade > b.grade;
    return a.name < b.name;
}

int main() {
    std::vector<Student> students = {
        {"Alice", 85}, {"Bob", 92}, {"Charlie", 85}
    };

    std::sort(students.begin(), students.end(), cmp);

    for (const auto& student : students) {
        std::cout << student.name << " " << student.grade << std::endl;
    }

    return 0;
}

File: 13-sorting/5-custom-comparators/main.java
// Custom Comparator - Java Example

import java.util.*;

class Student {
    String name;
    int grade;
    Student(String name, int grade) {
        this.name = name;
        this.grade = grade;
    }
    public String toString() {
        return name + " " + grade;
    }
}

class CustomComparator {
    public static void main(String[] args) {
        List<Student> students = Arrays.asList(
            new Student("Alice", 85),
            new Student("Bob", 92),
            new Student("Charlie", 85)
        );

        students.sort(Comparator.comparingInt((Student s) -> -s.grade).thenComparing(s -> s.name));

        for (Student s : students) {
            System.out.println(s);
        }
    }
}
